packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -586,75 +590,73 @@ class TimelineFrame {
   bool get isReadyForTimeline {
     return uiEventFlow != null &&
         gpuEventFlow != null &&
-        _pipelineItemStartTime != null &&
-        _pipelineItemEndTime != null;
+        pipelineItemStartTime != null &&
+        pipelineItemEndTime != null;
   }
 
-  /// Pipeline item start time in micros.
-  ///
-  /// This stores the timestamp of the start pipeline event for this frame. We
-  /// use this value to determine whether a TimelineEvent fits within the
-  /// frame's time boundaries.
-  int get pipelineItemStartTime => _pipelineItemStartTime;
-  int _pipelineItemStartTime;
-  set pipelineItemStartTime(int time) =>
-      _pipelineItemStartTime = nullSafeMin(_pipelineItemStartTime, time);
-
-  /// Pipeline item end time in micros.
-  ///
-  /// This stores the timestamp of the end pipeline event for this frame. We use
-  /// this value to determine whether a TimelineEvent fits within the frame's
-  /// time boundaries.
-  int get pipelineItemEndTime => _pipelineItemEndTime;
-  int _pipelineItemEndTime;
-  set pipelineItemEndTime(int time) =>
-      _pipelineItemEndTime = nullSafeMax(_pipelineItemEndTime, time);
+  // Stores frame start time, end time, and duration.
+  final timeRange = TimeRange();
 
-  bool get isWellFormed =>
-      _pipelineItemStartTime != null && _pipelineItemEndTime != null;
+  /// Pipeline item time range in micros.
+  ///
+  /// This stores the start and end times for the pipeline item event for this
+  /// frame. We use this value to determine whether a TimelineEvent fits within
+  /// the frame's time boundaries.
+  final _pipelineItemTimeRange = TimeRange();
 
-  /// Frame start time in micros.
-  int get startTime => _uiStartTime;
+  int get pipelineItemStartTime => _pipelineItemTimeRange.start?.inMicroseconds;
 
-  /// Frame end time in micros.
-  int get endTime => _gpuEndTime;
+  set pipelineItemStartTime(int time) => _pipelineItemTimeRange.start =
```

Comment:
I would suggest eliminating all pairs of members where the variables have 
StartTime or EndTime in the name. That is a sign the TimeRange class isn't being used fully yet.

Reply:
done.